### ISSUE TITLE
About Repository

### ISSUE THREAD

**Issue by Rokt33r** [BoostIO/boost-app-discussions#29]
## Current

Boostnote is now saving articles to a single JSON file.
The file can be located here

OS X
```
/Users/$USER_NAME$/Library/Application Support/boost/local.json
```

Windows
```
C:\Users\$USER_NAME$\AppData\Roaming\boost\local.json
```

I know it is not cool.

This is because the current app is just for checking Product Market Fit.
And, I wanted to focus on UI/UX matters. (also because of my programming skill.. :crying_cat_face:)

## Future

Anyway, It's about the time to make Cloud.
Here comes the future plan of Boostnote about Repository.

First, We won't offer our own storage server.
There are some reasons about it.

1. We have only one developer(Me: @rokt33r).
2. C2C note service(like Evernote) doesn't seem to be cool as a business although I also like my app.<sup>http://www.businessinsider.in/Evernote-the-first-dead-unicorn/articleshow/48989366.cms</sup>

Instead of dedicated cloud storage, We offer you another option.
Every article in Boostnote will be saved as a single file. So it can be saved on Github/BitBucket Repository.

### How?

Boostnote will exploit user's git client. The app will check whether the user have a git.
If user doesn't have it, We will offer installing instruction.(I think it is not so hard to install on any platform. it just like `next->next->...->install->complete`)

To be specific.

1. User create a repository and set URL of remote repo.
2. User update a note.
3. Boostnote commits the change and push it to remote. 
It will be done automatically by default. And also configurable to push it manually like `git add`.
4. The remote send a message to Boostnote webhook server after pushed.
5. Boostnote webhook server informs all Boostnote client app which is set same remote URL by Websocket.
6. Every client app pull the changes and automatically resolve the conflicts.

So, You can save your money. Also, You don't have to build up your own storage server.

Anyway I think it will be shipped on Feb. Then, I'll update this issue continuously.
Any opinions are welcome. Please feel free to comment here or make a new issue! :dancer:

**Comment by nmarus** [comment]
Great app! I love the code-centric direction and the ability to use markdown. I'm looking forward to your plans for other storage options. I am NOT looking forward to either writing a tool or manually moving nearly 500 notes out of ever note though... It's not a bad problem to have for a superior product. :) 

I think another cool idea would be the ability to point which folder the article(s) are stored locally would be great. Right now, I'm just doing sym link to Dropbox which works well for now. 

In addition to github repo as an option, it would nice to have the ability to selectively specify notes as github gists. 

If these are pay options, I'd gladly pay for this. Again, great work!

**Comment by Rokt33r** [comment]
@nmarus
Wow.... symlink... you are a genius!

Btw, Thanks for your comment! And, these are my answer!

### Importing from Evernote

I want to make it. But I don't know which extension would be available. I think We need to talk about it more!

### Syncing Gist

Sorry, It was planned but delayed... I'll make it possible soon!

### Pay option

I will provide all stuff free as written in Homepage. :tada:  Also, this app will become Opensource. About opensourcing, I'll make a new issue soon.

**Comment by isti37** [comment]
Don't make it cloud, integrate with git and add some encryption possibility and that would be even better.

**Comment by Rokt33r** [comment]
@isti37 
Of course! But, I don't know what can I do for 'some encryption possibility'. Could explain more about it?

**Comment by Rokt33r** [comment]
**Saving each post as an single file** will be released at the end of January. Although my final exam is going to begin soon... :smiling_imp:

**Comment by nmarus** [comment]
When you move to single file saves what will happen to data in local.json? Will that get migrated to individual files?

**Comment by Rokt33r** [comment]
@nmarus 
Boostnote will show some kind of a button to migrate. So, user can choose when to migrate.
Of course, the old repository would become obsolete somewhen(maybe about 2 months later after updating), but I don't want make people hurry.

**Comment by samrocksc** [comment]
Do you have a reasonable gathering of public feedback as to how they are using it?

I use it as a general purpose word processor, and note taker because of the awesome markdown integration.

I would be interested to know what direction you are seeing the most interested in boost with.

**Comment by Rokt33r** [comment]
@samrocksc 
Sorry, I can't get the point. Which thing is related to 'About Repository'?(This is not blaming. I just make sure that I understand what you say because I have poor English skill)

If your questions are general things, these are my answers.
<dl>
<dt>Do you have a reasonable gathering of public feedback as to how they are using it?</dt>
<dd>We're currently doing Offline interview(in Tokyo, Japan) and checking Twitter and this issue tracker.</dd>
<dt>I would be interested to know what direction you are seeing the most interested in boost with.</dt>
<dd>
Our goal is a replacement of Evernote. Which must be code-centric, hackable and giving free to choose storage.</dd>
</dl>

And, thank you very much for the praise about markdown integration. But, I think it still have a long way to go. If you have any idea and inconvenience, feel free to tell me. It would help me to make better Boostnote as your word processor and note taker.

**Comment by Rokt33r** [comment]
I'm so sorry. I think I have to postpone 'Saving each post as an single file' for about 2 weeks.
I thought I can finish it before my semester exams beginning. But, checking the current progress, it is impossible to release stable build with File things of Boostnote.
I'll resume my work as soon as the exams are over on next Thursday.

**Comment by alexanderadam** [comment]
@Rokt33r I am super excited to hear that the app will be open sourced at one point.
Can you roughly say when this will happen (or ping me when the day has come)? :wink: 

Thank you for your work!

**Comment by Rokt33r** [comment]
@alexanderadam 
It would be on this spring. :)

**Comment by javihernandez** [comment]
> It would be on this spring. :)

Let me know when it's ready too, I'd love to give it a try on GNU/Linux and also helping a bit on this task if needed

**Comment by Rokt33r** [comment]
@javihernandez
Of course, I'll! :dancer: 
Btw, I am going to buy new laptop to develop Linux app(or buy another MacBook and install Ubuntu to my old MacBook).
Could you recommend some laptops for me?

**Comment by javihernandez** [comment]
> Of course, I'll!  

Thanks!

> Btw, I am going to buy new laptop to develop Linux app(or buy another MacBook and install Ubuntu to my old MacBook).
> Could you recommend some laptops for me?

You can use any GNU/Linux distribution of your choice inside a VM. Lenovo and Dell are good options too if you want a dedicated linux laptop to develop with. :wink:

**Comment by Rokt33r** [comment]
@javihernandez 
Thanks for your comments! I will check Lenovo and Dell because I think I need dedicated one. :smiley:

**Comment by ericop** [comment]
Backup to Gist would be sweet! :candy: Good works Boost team!

**Comment by themightychris** [comment]
@Rokt33r check out how [GitBook Editor](https://www.gitbook.com/editor) formats its storage and auto-commits to git, I would LOVE it if something similar to that was Boost's default storage format/workflow.

I think it would be best to focus on first-class git integration and get everything we can out of that and not let supporting fast-and-loose schemes like dropbox/drive/evernote be a distraction or water down functionality. The stated goal is to be the note app for programmers so let's not let non-programmer note apps bring in a least common denominator.

**Comment by Rokt33r** [comment]
@themightychris Thank you. your comment is so helpful...! I've been wandering what should I do to implement it... I'll give a shot on this week. I really thank you again.

**Comment by themightychris** [comment]
@Rokt33r one of the qualities of GitBook's format that I think is worth striving for is maximizing how usable the notes are just through Github's UI.

For example, Gitbooks' SUMMARY.md, used for table of contents, is just a markdown list of links. When you look at it on Github you can just click links. Each tag in b00st could just be a .md file in a tags directory with a list of links to note files

**Comment by Rokt33r** [comment]
@themightychris 
What a cleaver solution! Generating summary of repository must be a great feature.

Btw, how can I keep the note stats, such as tags, created time, updated time?
Is it okay to generate some kind of manifest file as a json?

**Comment by themightychris** [comment]
@Rokt33r No need for your own metadata, git can do all that for you! If each note is its own file, git can tell you when it was created and last updated via its commit history.

Get the commit that added a file: `git log --diff-filter=A -- foo.js`
Get last commit that modified a file: `git log -n 1 -- README.md`

For tags just have one markdown file per tag in a tags/ directory with lists of links and boost can parse those and match up notes.

**Comment by Rokt33r** [comment]
@themightychris 
Sounds great. I'll try it on this week. I think I can finish the prototype version of it about next week.
I'll call you again after I finished. I really appreciate your help. :smile:

**Comment by themightychris** [comment]
@Rokt33r is there a branch somewhere I might be able to play with where you started toying with implementing this? It doesn't look like the current release or master branch incorporate any git integration for storage
